### PR TITLE
Add PREALLOCATE_SRV_MIP_CLAMPS workaround for Halo Infinite

### PR DESCRIPTION
**Pull request by esullivan-nvidia**
Halo Infinite currently has a race condition that can result in a GPU fault on NVIDIA hardware. The root cause of this race condition is the application overwriting existing SRV descriptors with a new mip clamp value asynchronously to the applications GPU submission timeline. This is problematic for vkd3d because it requires creating a new VkImageView to represent the SRV with the updated mip clamp value. If this SRV is being created after work has been submitted that accesses it, the GPU may not see the correct descriptor state possibly resulting in a fault.

This commit adds the VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS configuration as a workaround to the race condition. It works by preallocating all the VkImageViews for each possible mip clamp level the application may attempt to create for the resource. Technically this is still a race condition but the end result isn't a GPU fault if the race condition is lost. Instead the shader will just sample from the texture with the previous SRVs mip clamp level rather than the updated one. These are the same outcomes that are possible on RADV without this workaround.

### CODE REVIEW

File: libs/vkd3d/resource.c [hunk 1]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
     if (key.u.texture.miplevel_count == VK_REMAINING_MIP_LEVELS)
         key.u.texture.miplevel_count = resource->desc.MipLevels - key.u.texture.miplevel_idx;
 
+    if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS) &&
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMS) &&
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMSARRAY) &&
+        (hash_map_find(&resource->view_map.map, &key) == NULL))
+    {
+        uint32_t starting_mip = key.u.texture.miplevel_idx;
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4440]: Starting mip should be +1 here since we create the base level below.

File: libs/vkd3d/resource.c [hunk 2]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
...
+        uint32_t mip_count = key.u.texture.miplevel_count != UINT32_MAX ?
+            key.u.texture.miplevel_count :
+            resource->desc.MipLevels - starting_mip;
+
+        struct vkd3d_view_key preallocate_key = key;
+
+        for (uint32_t i = starting_mip; i < mip_count; i++)
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4447]: We use C89 style for the loop initializer, so it should be done outside the loop.

File: libs/vkd3d/resource.c [hunk 3]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
     if (key.u.texture.miplevel_count == VK_REMAINING_MIP_LEVELS)
         key.u.texture.miplevel_count = resource->desc.MipLevels - key.u.texture.miplevel_idx;
 
+    if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS) &&
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMS) &&
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMSARRAY) &&
+        (hash_map_find(&resource->view_map.map, &key) == NULL))
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4438]: !hash_map_find(...) is more in-style.

File: libs/vkd3d/resource.c [hunk 4]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
     if (key.u.texture.miplevel_count == VK_REMAINING_MIP_LEVELS)
         key.u.texture.miplevel_count = resource->desc.MipLevels - key.u.texture.miplevel_idx;
 
+    if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS) &&
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMS) &&
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4436]: Can drop the parens here. and the one below.

File: libs/vkd3d/resource.c [hunk 5]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
...
+        (desc->ViewDimension != D3D12_SRV_DIMENSION_TEXTURE2DMSARRAY) &&
+        (hash_map_find(&resource->view_map.map, &key) == NULL))
+    {
+        uint32_t starting_mip = key.u.texture.miplevel_idx;
+        uint32_t mip_count = key.u.texture.miplevel_count != UINT32_MAX ?
+            key.u.texture.miplevel_count :
+            resource->desc.MipLevels - starting_mip;
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4443]: I think this is wrong. starting_mip is relative to the resource itself and we're iterating from starting_mip to MipLevels here.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Some nits.

File: libs/vkd3d/resource.c [hunk 6]
```diff
@@ -4432,6 +4432,25 @@ static void vkd3d_create_texture_srv(vkd3d_cpu_descriptor_va_t desc_va,
     if (key.u.texture.miplevel_count == VK_REMAINING_MIP_LEVELS)
         key.u.texture.miplevel_count = resource->desc.MipLevels - key.u.texture.miplevel_idx;
 
+    if ((vkd3d_config_flags & VKD3D_CONFIG_FLAG_PREALLOCATE_SRV_MIP_CLAMPS) &&
```
Comment (HansKristian-Work) [inline libs/vkd3d/resource.c:4435]: Given that we only add clamps for integer mips, we should probably floor() the mip clamp when this workaround is enabled.

File: libs/vkd3d/resource.c (hunk 1 above)
Comment (esullivan-nvidia) [reply libs/vkd3d/resource.c:4440]: I don't think this code can assume the application will always initially create the SRV with the mip clamp set to the same level as MostDetailedMip. Any value in the range [MostDetailedMip, MostDetailedMip + MipLevels] would be valid.

File: libs/vkd3d/resource.c (hunk 2 above)
Comment (esullivan-nvidia) [reply libs/vkd3d/resource.c]: Fixed

File: libs/vkd3d/resource.c (hunk 3 above)
Comment (esullivan-nvidia) [reply libs/vkd3d/resource.c]: Fixed

File: libs/vkd3d/resource.c (hunk 4 above)
Comment (esullivan-nvidia) [reply libs/vkd3d/resource.c]: Fixed

**Comment by esullivan-nvidia** [comment]
> This is truly cursed. This must have been a true nightmare to debug. Thanks a lot for this. Can you confirm this contribution is done on behalf of NVIDIA given your commit email (w.r.t the MIT relicense process we're working through)?

Yes I can. This contribution is done on behalf of NVIDIA.

**Comment by esullivan-nvidia** [comment]
> FWIW, this is clearly a game bug. It's not allowed to do this in D3D12 either, has the bug been reported to the developers already?

Yes, it has been reported to the developer.

File: libs/vkd3d/resource.c (hunk 1 above)
Comment (HansKristian-Work) [reply libs/vkd3d/resource.c]: Fair enough.

**Pull request merged**